File: src/main/java/no/nav/permitteringsskjemaapi/config/KafkaConfig.java
package no.nav.permitteringsskjemaapi.config;

import org.apache.kafka.clients.producer.ProducerConfig;
import org.apache.kafka.common.config.SslConfigs;
import org.apache.kafka.common.serialization.StringSerializer;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.kafka.core.DefaultKafkaProducerFactory;
import org.springframework.kafka.core.KafkaTemplate;
import org.apache.kafka.clients.CommonClientConfigs;

import java.util.HashMap;
import java.util.Map;

@Configuration
public class KafkaConfig {
    private final String keystorePath;
    private final String credstorePassword;
    private final String truststorePath;

    public KafkaConfig(
            @Value("${kafka.keystore.path}") String keystorePath,
            @Value("${kafka.credstore.password}") String credstorePassword,
            @Value("${kafka.truststore.path}")String truststorePath)
             {
        this.keystorePath = keystorePath;
        this.credstorePassword = credstorePassword;
        this.truststorePath = truststorePath;
    }

    @Bean
    public KafkaTemplate kafkaTemplate() {
        Map<String, String> configMap = new HashMap<>();
        configMap.put(SslConfigs.SSL_KEYSTORE_LOCATION_CONFIG,keystorePath);
        configMap.put(SslConfigs.SSL_TRUSTSTORE_PASSWORD_CONFIG,credstorePassword);
        configMap.put(SslConfigs.SSL_KEYSTORE_PASSWORD_CONFIG,credstorePassword);
        configMap.put(SslConfigs.SSL_KEY_PASSWORD_CONFIG,credstorePassword);
        configMap.put(SslConfigs.SSL_TRUSTSTORE_TYPE_CONFIG,"JKS");
        configMap.put(SslConfigs.SSL_KEYSTORE_TYPE_CONFIG,"PKCS12");
        configMap.put(SslConfigs.SSL_TRUSTSTORE_LOCATION_CONFIG,truststorePath);
        configMap.put(CommonClientConfigs.SECURITY_PROTOCOL_CONFIG,"SSL");
        configMap.put(ProducerConfig.CLIENT_ID_CONFIG,"permitteringsskjema-api");
        configMap.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName());
        configMap.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName());

        return new KafkaTemplate(new DefaultKafkaProducerFactory(configMap));
    }
}

File: src/main/kotlin/no/nav/permitteringsskjemaapi/config/KafkaConfig.kt
package no.nav.permitteringsskjemaapi.config

import org.apache.kafka.clients.consumer.ConsumerRecord
import org.apache.kafka.clients.producer.ProducerConfig
import org.apache.kafka.common.serialization.StringSerializer
import org.springframework.boot.autoconfigure.kafka.KafkaProperties
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.context.annotation.Profile
import org.springframework.kafka.annotation.EnableKafka
import org.springframework.kafka.config.ConcurrentKafkaListenerContainerFactory
import org.springframework.kafka.core.DefaultKafkaConsumerFactory
import org.springframework.kafka.core.DefaultKafkaProducerFactory
import org.springframework.kafka.core.KafkaTemplate
import org.springframework.kafka.listener.DefaultErrorHandler
import org.springframework.kafka.listener.RetryListener
import org.springframework.util.backoff.ExponentialBackOff

@Configuration
@EnableKafka
class KafkaConfig {
    private val log = logger()

    @Profile("dev-gcp", "prod-gcp")
    @Bean
    fun errorLoggingKafkaListenerContainerFactory(
        properties: KafkaProperties
    ): ConcurrentKafkaListenerContainerFactory<String, String> {
        val factory = ConcurrentKafkaListenerContainerFactory<String, String>()
        factory.consumerFactory = DefaultKafkaConsumerFactory(properties.buildConsumerProperties())
        val errorHandler = DefaultErrorHandler(ExponentialBackOff())
        errorHandler.setRetryListeners(
            RetryListener { r: ConsumerRecord<*, *>, ex: Exception, attempt: Int ->
                log.error(
                    "KafkaListener failed. attempt={} topic={} parition={} offset={} key={}  exception={}",
                    attempt, r.topic(), r.partition(), r.offset(), r.key(), ex.javaClass.canonicalName, ex
                )
            }
        )
        factory.setCommonErrorHandler(errorHandler)
        return factory
    }

    @Bean
    fun kafkaTemplate(
        producerFactory: DefaultKafkaProducerFactory<String, String>
    ) = KafkaTemplate(
        producerFactory, mapOf(
            ProducerConfig.CLIENT_ID_CONFIG to "permitteringsskjema-api",
            ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG to StringSerializer::class.java.name,
            ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG to StringSerializer::class.java.name,
        )
    )
}